src/go/k8s/pkg/resources/statefulset.go

```diff
@@ -372,15 +399,57 @@ func (r *StatefulSetResource) Kind() string {
 func (r *StatefulSetResource) portsConfiguration() string {
 	kafkaAPIPort := r.pandaCluster.Spec.Configuration.KafkaAPI.Port
 	rpcAPIPort := r.pandaCluster.Spec.Configuration.RPCServer.Port
-	svcName := r.svc.Key().Name
-
+	svcName := r.headlessSvc.Key().Name
+	externalKafkaAPIPort := r.pandaCluster.Spec.Configuration.KafkaAPI.Port + 1
+
+	if r.pandaCluster.Spec.ExternalConnectivity {
+		return fmt.Sprintf("--kafka-addr=internal://$(POD_IP):%d"+
+			" --advertise-rpc-addr=$(POD_NAME).%s.$(POD_NAMESPACE).svc.cluster.local.:%d"+
+			" --rpc-addr=$(POD_IP):%d"+
+			" --kafka-addr=external://$(POD_IP):%d",
+			kafkaAPIPort,
+			svcName, rpcAPIPort,
+			rpcAPIPort,
+			externalKafkaAPIPort)
+	}
 	// In every dns name there is trailing dot to query absolute path
 	// For trailing dot explanation please visit http://www.dns-sd.org/trailingdotsindomainnames.html
-	return fmt.Sprintf("--advertise-kafka-addr=internal://$(POD_NAME).%s.$(POD_NAMESPACE).svc.cluster.local.:%d"+
-		" --kafka-addr=internal://$(POD_IP):%d"+
+	return fmt.Sprintf("--kafka-addr=internal://$(POD_IP):%d"+
 		" --advertise-rpc-addr=$(POD_NAME).%s.$(POD_NAMESPACE).svc.cluster.local.:%d"+
 		" --rpc-addr=$(POD_IP):%d",
-		svcName, kafkaAPIPort, kafkaAPIPort, svcName, rpcAPIPort, rpcAPIPort)
+		kafkaAPIPort,
+		svcName, rpcAPIPort,
+		rpcAPIPort)
+}
+
+func (r *StatefulSetResource) getPorts() []corev1.ContainerPort {
+	if r.pandaCluster.Spec.ExternalConnectivity &&
+		r.nodePortSvc.LastObservedState != nil &&
```

Comment:
I don't like this coupling, I would rather explicitly depend on list of ports that needs to be added here. Let's also keep the logic that the ports are kafkaapi+1 in one place so ideally this should already get list of ports and just add it, not again duplicate the logic that it's kafka + 1

Reply:
>  I would rather explicitly depend on list of ports

It's not possible due to fact that each resource is created in its own `Ensure` function. The controller can not provide in the StatefulSetResource "constructor" the list of ports.

> Let's also keep the logic that the ports are kafkaapi+1 in one place

Agree I will address that.

Reply:
oh I see, I bet there's a nice way to get around that, just add this as (the reason) a comment and let's keep it like this

Reply:
I changed the if statement so there is no coupling right now.